Task - test coverage and lots of core InVEST UI functionality

davemfish opened the pull request:
The goals for this PR were to reproduce the existing functionality of the python InVEST UI, and to have good documentation and test coverage.

Mainly this is an oppurtunity to get up to speed with the latest development of this project, and discuss and any changes of course that we might want to make.  

**Some things I feel good about and are less likely to change dramatically:**
* ReactJS for building the UI components
* leveraging existing functionality of `natcap.invest` (`validation`, loading model `ARGS_SPEC`, `datastack` functionality). Right now this happens with a Flask interface and that works quite well. But I don't think we ever really considered socket communication. So something to consider if we find shortcomings in the Flask approach.
* a node subprocess call to the invest CLI for running an invest model
* Jest + react-testing-library & dom-testing-library for integration and unit tests

**Some things that are more likely to change:**
* Packaging & distribution. Right now the app only works in a development setting. Some work has been done with packaging (see readme), but we have not tackled the python side of this at all.
* Somewhat related to above - the tests currently do not touch the python side (flask server or invest CLI). This is mainly to maintain speed and reliability of the tests. But we shouldn't ignore test coverage for this.
* Using a json file on disk as a persistent data store. This allows quick re-load of past sessions in a similar way that the current UI populates with values from a previous session by default. You'll notice the `jobdb.json` that gets created and added to anytime you run a model. 
* There are lots of places throughout the code with `TODO` flags and comments, but they are all fairly specific. 
* Almost everything related to style and appearance
* Anything related to open issues, especially natcap/invest-workbench#1,  natcap/invest-workbench#6, natcap/invest#874, 

**Getting setup:**
The readme is a good entry point for getting things setup. 

As a reminder, electron applications have two mostly separate javascript processes. The main process and the render process. In this project, `npm start` runs `main.js` (the main electron process) in a node environment. This process creates a browser window and loads `index.html`, which includes `index.js` as a script. `index.js` is the starting point of the render process, and it renders the root react component, `app.jsx`.

Please don't bother reviewing anything underneath `src/components/ResultsTab/Visualization/`. These are the model-specific visualizations and they only exist for Carbon and HRA. The HRA component is quite large and was developed in a separate project.

This PR should resolve issues natcap/invest-workbench#5 natcap/invest-workbench#3 natcap/invest-workbench#9 natcap/invest#876

davemfish committed:
installed enzyme for DOM manipulation and testing HomeTab interactions. #5.

davemfish committed:
added test for a click on a recent session button. #5.

davemfish committed:
refactored a server request into another module, to facilitate mocking. #5.

davemfish committed:
added an integration test using react-testing-library. #5.

davemfish committed:
replacing node requests library with node-fetch. requests is deprecated. #5

davemfish committed:
finished removing the requests library in favor of node-fetch.

davemfish committed:
uninstalled enzyme. #5

davemfish committed:
several minor package updates. #5.

davemfish committed:
some jest config to see a coverage report. #5.

davemfish committed:
working integration test for clicking a model and loading the SetupTab. #5.

davemfish committed:
added assertion on button property. #5.

davemfish committed:
fixing the check for a disabled button. #5.

davemfish committed:
cleaned some commented code. #5.

davemfish committed:
added a unique key to a component in a list. #5.

davemfish committed:
added a test for rendering setup after clicking a recent session button. #5.

davemfish committed:
disabling save parameter options before any model is selected, and testing for that. #5.

davemfish committed:
select node should have a defaultValue attribute instead of a value attribute. #5.

davemfish committed:
working on test for settings dialog behavior. #5.

davemfish committed:
completed test of settings dialog in new app.test file. #5.

davemfish committed:
skipping tests that still depend on enzyme - these need refactoring. #5.

davemfish committed:
install and configured jest to run in an electron environment instead of node/jsdom. #5.

davemfish committed:
removing the electron jest-runner from the jest config as it makes debugging challenging. #5.

davemfish committed:
extending testing-library expectations with jest-dom and trying to use them where possible. #5.

davemfish committed:
added a right-click menu with Inspect Element. #5.

davemfish committed:
correcting the validation mock and subsequent Disabled checks on Execute button. #5.

davemfish committed:
changing dropdown items to buttons so they can carry a disabled attribute. #5.

davemfish committed:
minor refactor of the save dropdown components. #5.

davemfish committed:
in progress refactor of streaming logfiles to display.

davemfish committed:
adding dotenv dependency for setting environment variables.

davemfish committed:
added utils file and function for finding the most recent logfile from a workspace.

davemfish committed:
using LazyLog for streaming a logfile. in progress.

davemfish committed:
trying Tail instead of LazyLog. in progress.

davemfish committed:
using tail to render log from the contents of the invest logfile.

davemfish committed:
fixed a bug where execute was firing before the datastack file it needed was written.

davemfish committed:
automatic save state on invest execute.

davemfish committed:
added support for metadata on jobs, parsed at the app-level into the recent sessions buttons.

davemfish committed:
storing and loading persistent invest job metadata with a json file.

davemfish committed:
getting tests passing with refactored recentSessions structure.

davemfish committed:
pass persistent jobs database as a prop when rendering App. Also a test expecting a button for each recent job. #5.

davemfish committed:
tests for the SetupTab arguments form. in progress. #5.

davemfish committed:
basic assertions on correct form input types.

davemfish committed:
debugging setup tab input form tests. #5.

davemfish committed:
working tests for individual inputs, and a new test for the whole of an arguments form. #5.

davemfish committed:
a not-working attempt to test investExecute by mocking child_process.spawn. #5.

davemfish committed:
removed the non-functional method for killing the spawned invest subprocess.

davemfish committed:
added tests for the log display. #5.

davemfish committed:
testing for an edge case where non-fatal stderr is emmitted from invest. fixes #3. addresses #5.

davemfish committed:
adding this file that I forgot to add along with app.test.js. #5.

davemfish committed:
remove an obsolete test file. #5.

davemfish committed:
removing an obsolete snapshot file. #5.

davemfish committed:
installed spectron for end-to-end tests and attempting to use it. in progress. #5.

davemfish committed:
upgraded to electron version 8. #5.

davemfish committed:
successfully starting and stopping the app in a test, using spectron. #5

davemfish committed:
minor changes.

davemfish committed:
setting up to use electron forge to package the app.

davemfish committed:
using esm for module loading instead of babel/register.

davemfish committed:
checking for dev vs production before loading some modules.

davemfish committed:
use babel cli to transpile jsx.

davemfish committed:
packaging with electron-builder instead of electron-forge

davemfish committed:
use an env var to determine if we're in dev mode in the renderer process, and whether or not babel/register needs to be required.

davemfish committed:
debugging some transpile and packaging stuff.

davemfish committed:
adding a custom build script for preparing dev code for production packaging.

davemfish committed:
cleaning up some comments

davemfish committed:
trying out some props type-checking with prop-types. #5.

davemfish committed:
added props type-checking for all components. #5.

davemfish committed:
unit tests for find invest logfile utility. #5.

davemfish committed:
unit tests for updateRecentSessions. #5.

davemfish committed:
documentation for InvestJob methods.

davemfish committed:
more unit tests for some utility functions. also moved some functions to utils.js. #5.

davemfish committed:
adding more documentation. #5.

davemfish committed:
removed the feature to browse to and load an invest job state from a file.

davemfish committed:
removed an extranneous wrapper function. #5.

davemfish committed:
removing a flask endpoint that was a bad idea to begin with.

davemfish committed:
more documentation.

davemfish committed:
more comments.

davemfish committed:
removed the feature to manually save the current session at any time.

davemfish committed:
removing the import for this deprecated feature.

davemfish committed:
documenting some more components. #5.

davemfish committed:
fix the regex for finding invest logfiles that are generated by the invest CLI. #5.

davemfish committed:
some style adjustments for recent session cards.

davemfish committed:
seems the electron API changed and showOpenDialog now returns a promise

davemfish committed:
added component for loading params from a file. this includes a small refactor of investGetSpec to return a promise.

davemfish committed:
fixing regex to handle another character in an invest logfile name.

davemfish committed:
reverting the recent change in getspec endpoint logic, now preparing the model name beforehand as needed.

davemfish committed:
testing the load parameters button. includes checking in a new manual mock for electron.remote. #5.

davemfish committed:
added tests for the save dropdown buttons. #5.

davemfish committed:
tests for save dialogs being cancelled. #.5

davemfish committed:
testing the browse button functionality on the SetupTab. #5.

davemfish committed:
added a test for drag-drop loading of datastacks/logfiles. #5.

davemfish committed:
added support for showing About text for each invest argument.

davemfish committed:
tests for the about text dialog boxes. #5.

davemfish committed:
added button to open the output invest workspace upon model completion. #5.

davemfish committed:
manually adjusting the project's version number to reflect current state of development. #5.

davemfish committed:
removing an obsolete snapshot

davemfish committed:
putting some placeholder tests in these files rather than untracking the files. #5.

davemfish committed:
fixing a bug in the filepath of saved state files.

davemfish committed:
checkin a file used as test data. #5.

davemfish committed:
new method for checking if the flask server is ready, and waiting until it is. #5.

davemfish committed:
removing a comment.

davemfish committed:
fixing sample path to statefile in the sample jobs database.

davemfish committed:
handling an edge case for a missing logfile.

davemfish committed:
some updates to comments.

davemfish committed:
some logging for an intermittent bug.

davemfish committed:
dealing with an intermittent bug in module loading by explictly loading an .mjs file instead of a .js file.

davemfish committed:
renamed file

davemfish committed:
some comments

davemfish committed:
debugging getspec calls that start with a datastack as opposed to an invest list model name.

davemfish committed:
removing an es6 module dependency from main.js to simplify imports.

davemfish committed:
updated node-fetch package.

davemfish committed:
some logging related to an intermittent bug

davemfish committed:
untracking the incomplete coastal vulnerability visualization.

davemfish committed:
ignoring the job database json file.

davemfish committed:
handle cases where recent session data is missing, instead of just crashing.

davemfish committed:
reducing the use of one of the sessionProgress variable so that it only has one real purpose. also some comments.

davemfish committed:
some developer notes.

davemfish committed:
changing some temporary text so as not to break a DOM query in tests.

davemfish committed:
changing some temporary text so as not to break a DOM query in tests.

davemfish committed:
updating my ignore file.

davemfish committed:
removing electron-forge config from package.json since we're not currenty using electron-forge.

davemfish committed:
updates to readme

davemfish committed:
quick bugfix for a missing import in main.js.

davemfish committed:
halt recursive calls that find out if the flask server is ready, after n tries.

phargogh commented on src/server.py:
```diff
@@ -44,10 +42,13 @@
     "sdr": "sdr.sdr",
     "seasonal_water_yield": "seasonal_water_yield.seasonal_water_yield",
     "urban_flood_risk_mitigation": "urban_flood_risk_mitigation",
+    "urban_cooling_model": "urban_cooling_model",
```
I could be wrong about this, but this reminds me a lot of the `_MODEL_UIS` (or some name like that) dict in the CLI file, and it could make sense to reduce the duplication here since this is on the python side of things.  Would that make sense for this case?

davemfish replied:
Yes, I think that would work!

phargogh commented on tests/app.test.js:
```diff
@@ -0,0 +1,93 @@
+import fs from 'fs';
+import React from 'react';
+import { fireEvent, render, wait } from '@testing-library/react'
+import '@testing-library/jest-dom'
```
What's the significance of the leading `@`?

davemfish replied:
Good question. My reading is, `jest-dom` is the package, and it's developed under the scope of `@testing-library` project. Mainly that "scoping" has implications for npm. There's no other way to do the import here.

https://stackoverflow.com/questions/36667258/what-is-the-meaning-of-the-at-prefix-on-npm-packages

phargogh replied:
Ah!  That is a really interesting feature, and a bit surprising that it's implemented with some different import syntax.

phargogh reviewed (approved):
This is really exciting @davemfish !  I don't think I have any specific requests for things to change, just added a few comments as I had questions reading through things.  Overall, I think the project organization makes sense, I really like how the application and test logic is expressed throughout, and I can see how we might expand it down the road as needed!  I'm sure we'll trend towards some sort of preference or standard for things like imports, docstrings (or whatever the js equivalent is), etc, and I'd defer to you on what would be best.

Anyways, I added by 'approval' here, and very happy to talk more!

dcdenu4 commented on .env-example:
```diff
@@ -1,10 +1,27 @@
+# It's best to define the invest executeable path here, having it on your 
+# system's path already might work, but might not.
+
+INVEST="C:\\Users\\dmf\\Miniconda3\\envs\\invest-py37x\\Scripts\\invest.exe"
```
For this file and line particularly, will this somehow be set by wherever the user Downloads InVEST to? Just curious how this file might look for packaging and deployment.

davemfish replied:
Good question, and part of larger discussion related to #1 . 

Right now, the `.env` file actually only loads in dev mode (i.e. the app is launched via `npm start`), not production mode. But I think it's a valid solution for production as well -- and we have no other solution for this yet, So we could generate and package a `.env` like this with the correct parameters set, depending on OS we're building for.

dcdenu4 commented on .gitignore:
```diff
@@ -1,2 +1,8 @@
+build/
```
I know there is much debate about including `.gitignore` in a repository or not. I'm fine with it, just wanted to make sure it was intended.

davemfish replied:
I don't feel strongly either. I probably tracked it for my own convenience of developing on different machines. But this might be a good time to untrack it.

dcdenu4 commented on readme.md:
```diff
@@ -4,8 +4,26 @@
 
```
Does it make sense to have a little blurb at the top of the readme just generally describing what this project is and it's purpose? I realize no one will probably be finding themselves here for awhile but, might be nice.

davemfish replied:
Yes, never hurts! I'll put this on my list for updating in this PR.

dcdenu4 commented on src/InvestJob.jsx:
```diff
@@ -38,32 +39,48 @@ if (process.env.GDAL_DATA) {
+      activeTab: 'home'                // controls which tab is currently visible
     };
     
     this.argsToJsonFile = this.argsToJsonFile.bind(this);
```
Hey Dave, could you make a comment about what these bind calls are doing? Not necessarily a code comment, but just a comment here for my curiosity. Do they help in handling callbacks or something?

dcdenu4 commented on src/InvestJob.jsx:
```diff
@@ -73,59 +90,68 @@ export class InvestJob extends React.Component {
 
-  saveState(event) {
-    // Save a snapshot of this component's state to a JSON file.
+  saveState() {
```
Ha, oh I think I get it relating to the bind comment.

davemfish replied:
Oh good! I'll explain it for myself anyway, just to convince myself that I also understand it. The `saveState` method wants to access properties of it's parent component (`InvestJob`). And so the method needs to be bound to that component. Once bound, `saveState`'s `this` object is also `InvestJob`'s `this`, so we can call `this.setState()` from inside the method, for example.

If not bound, `this` inside of `saveState` will not refer to the parent component's `this` at all.

dcdenu4 commented on src/InvestJob.jsx:
```diff
@@ -136,261 +162,272 @@ export class InvestJob extends React.Component {
-        directory: this.state.args.workspace_dir.value,
-        suffix: this.state.args.results_suffix.value
-      }
+      // TODO: there are non-zero exit cases that should be handled
+      // differently from one-another, but right now they are all exit code 1.
+      // E.g. this callback is designed with a model crash in mind, but not a fail to 
+      // launch, in which case the saveState call will probably crash.
+      const status = (code === 0 ? 'success' : 'error')
```
This is interesting.

davemfish replied:
I made issue #18 to track this. I'll need to do some more testing to think about a good way forward. One problem which I sometimes hit during dev is an error from the shell -- not even from the CLI -- if `spawn` was given a bad `cmd`, for example.

dcdenu4 commented on src/index.js:
```diff
@@ -10,13 +12,34 @@ var _reactDom = _interopRequireDefault(require("react-dom"));
+const menu = new Menu();
+menu.append(new MenuItem({
+  label: 'Inspect Element',
+  click: () => { 
+    remote.getCurrentWindow().inspectElement(rightClickPosition.x, rightClickPosition.y)
+  }
+}))
```
Should something here have a semicolon? Actually above in lines 20, 21 there also are no semicolons.

davemfish replied:
And I'm sure they're missing in many other places. I haven't been consistent. These days semicolons are added by the interpreter as needed, but there's still much debate. If we decide on a style guide we can go from there.

dcdenu4 commented on src/server.py:
```diff
@@ -57,17 +58,18 @@ def shutdown_server():
 
 @app.route('/ready', methods=['GET'])
 def get_is_ready():
-    return json.dumps('Flask ready')
+    return 'Flask ready'
```
Should these tiny functions have docstrings still?

davemfish replied:
Yes, will do.

davemfish commented on package.json:
```diff
@@ -1,87 +1,94 @@
   "name": "invest-electron",
   "productName": "invest-electron",
-  "version": "1.0.0",
+  "version": "0.1.0-alpha",
```
Yes! `npm version` is the tool for that. So a release pipeline/checklist could include commands like:

`npm version patch`
`npm version minor` etc,

These automatically update the package.json version string -- according to semver specs -- and commit a git tag with the same string.

phargogh replied:
Very cool! `npm version` looks really replete

davemfish commented on src/InvestJob.jsx:
```diff
@@ -38,32 +39,48 @@ if (process.env.GDAL_DATA) {
 const CACHE_DIR = 'cache' //  for storing state snapshot files
 const TEMP_DIR = 'tmp'  // for saving datastack json files prior to investExecute
 
+// to translate to the invest CLI's verbosity flag:
+const LOGLEVELMAP = {
+  'DEBUG':   '--debug',
+  'INFO':    '-vvv',
+  'WARNING': '-vv',
+  'ERROR':   '-v',
+}
+
```
I agree, I tend to prefer the more descriptive flags for command-line utilities. But I don't think it matters from the perspective of this UI.

davemfish commented on src/InvestJob.jsx:
```diff
@@ -73,59 +90,68 @@ export class InvestJob extends React.Component {
+      console.log("saved: " + jobName);
     });
-    this.props.updateRecentSessions(sessionID);
+    let job = {};
+    job[jobName] = {
+      model: this.state.modelName,
+      workspace: this.state.workspace,
+      statefile: filepath,
+      status: this.state.jobStatus,
+      humanTime: new Date().toLocaleString(),
+      systemTime: new Date().getTime(),
+    }
+    this.props.updateRecentSessions(job, this.props.appdata);
```
That's right. `updateRecentSessions` will overwrite an existing entry with an identical `sessionID`, and that's intended since invest will also overwrite that model run. I think this okay for now, but if someday invest caches some interesting data that is not over-written when a workspace + suffix are re-used, then we could revisit this.

I'm pretty sure we're going to revisit the structure of this job metadata and the fact that it's just stored in a json file anyway. This is one of the most experimental parts of the UI at this point.

davemfish replied:
Also in your example above: "what would happen if we re-run the same model in the same directory with the same suffix, but with different inputs?"

If a user did overwrite a previous run, but then wanted to redo that original run, they would no longer be able to get there by loading a "recent session", but they would still have the invest logfile from the earlier run and could load that via "Load Parameters" button.

davemfish commented on src/InvestJob.jsx:
```diff
@@ -136,261 +162,272 @@ export class InvestJob extends React.Component {
       model_module: this.state.modelSpec.module,
       args: args_dict_string
     };
+
+    // TODO: is there a use-case for `limit_to`? 
+    // Right now we're never calling validate with a limit_to,
+    // but we have an awful lot of logic here to cover it.
     if (limit_to) {
       payload['limit_to'] = limit_to
     }
```
Thus far this new UI presents no reason to keep `limit_to`. There is issue #17 - related to runtime of `validate` functions. But I doubt the solution there is `limit_to`.

davemfish commented on src/main.js:
```diff
@@ -81,17 +82,15 @@ function createPythonProcess() {
-  );
+function shutdownPythonProcess() {
+  return(
+    fetch('http://localhost:5000/shutdown', {
```
Seems like a good idea to me!

dcdenu4 commented on src/server_requests.js:
```diff
@@ -0,0 +1,126 @@
+  * so awaiting this response is one way to avoid that. 
+  */
+  return(
+    fetch('http://localhost:5000/ready', {
```
I think James mentioned this being configurable somehow?sa

dcdenu4 commented on tests/utils.test.js:
```diff
@@ -0,0 +1,133 @@
+import path from 'path';
+import { findMostRecentLogfile, loadRecentSessions,
+         updateRecentSessions, boolStringToBoolean,
+         argsValuesFromSpec } from '../src/utils';
```
I don't think `argsValuesFromSpec` is used anywhere?

dcdenu4 reviewed (approved):
Hey @davemfish this is some great work! And there's a lot of it. I put in some pretty small comments and questions. I think it will be great to have this officially tracked now and I'm looking forward to diving into pieces and parts as things come up. 

A few things I noticed:
- When selecting the 'X' close button in the app window the shell/terminal will hang on `Flask server shutting down...`
- Is the package-lock.json file something to be under version control?
- Seemed like there were some spots where semicolons weren't used, is there a JS best practice for semicolons? It's possible these spots I noticed don't actually need them, I get confused.

davemfish commented:
> 
> A few things I noticed:
> 
>     * When selecting the 'X' close button in the app window the shell/terminal will hang on `Flask server shutting down...`
> 
Yeah I notice that too. Flask is shut down, so I think it's the electron/node process that's still not terminating. I'll look into that.

>     * Is the package-lock.json file something to be under version control?
It is, yes, but it's certainly not something whose changes should be reviewed. And it shouldn't be edited manually. It's useful because it describes dependenices with exact versions that npm has already determined are compatible by solving the more general requirements defined in `package.json`. So when the lock file is present `npm install` will avoid solving the dependency tree if it can - like if there's no new deps in `package.json`. That said, if the env gets messed up, it's sometimes useful to delete the `package-lock.json` and do `npm install`, which then refers only to the `package.json` (and re-makes a new `package-lock.json` in the process).
> 
>     * Seemed like there were some spots where semicolons weren't used, is there a JS best practice for semicolons? It's possible these spots I noticed don't actually need them, I get confused.
I think it's up to us to decide whether we want to code them, or let the interpreter insert them as needed. I've been inconsistent thus far.

davemfish commented:
Thanks for reviewing this @phargogh and @dcdenu4 ! So far I responded to the conversations you both started, and I've got a running list of items that you suggested (generally the small things) that I'm going to update while this PR is open. And then I added issues for the larger things, like #18 and #1. And I might do #17 in this PR also because it's pretty high priority.

davemfish committed:
make port number for flask configureable in .env. #5.

davemfish committed:
replace a hardcoded lookup with one that already exists in natcap.invest.cli. #5.

davemfish committed:
added docstrings to server.py. #5.

davemfish committed:
untracking .gitignore. #5.

davemfish committed:
added a project description to the readme. #5.

davemfish committed:
added the same license as the invest license. #5.

davemfish committed:
just some linting. #5.

davemfish committed:
using datastack.py instead of js to write an invest datastack json. #5.

davemfish committed:
adding an explicit app.quit when all windows are closed. #5.

davemfish committed:
a workaround for the intermittent bug of not getting invest list response. #5.

davemfish commented:
Hey @phargogh and @dcdenu4 , I pushed changes with the following updates:

* configure port in .env
* replace the dict in server.py with cli._MODEL_UIS
* untrack .gitignore
* update Readme with general project desc/purpose
* docstrings in server.py
* added LICENSE from invest
* some linting
* properly terminating the app
* writeParamsToFile using datastack.py api, again. And test it.
* addressed the flaky not-fetching model list bug with a workaround

I think that's all I plan to do within this PR unless I hear otherwise from you!

davemfish commented on src/InvestJob.jsx:
```diff
@@ -136,261 +162,272 @@ export class InvestJob extends React.Component {
+    
+    // TODO: is there much to be gained by outsourcing this to datastack.py via flask?
+    // It would insert the invest version string and enable an option to pass
+    // relative_paths=True. This is what that looks like:
```
I switched this back to the `datastack.py` method in the latest changes. I think the reason I departed from that originally was to make testing easier, but I've gotten better at testing since then!

davemfish commented on src/InvestJob.jsx:
```diff
@@ -136,261 +162,272 @@ export class InvestJob extends React.Component {
-        logStdOut: ''
-      }
-    );
+    // Write a temporary datastack json for passing as a command-line arg
+    const temp_dir = fs.mkdtempSync(path.join(process.cwd(), TEMP_DIR, 'data-'))
+    const datastackPath = path.join(temp_dir, 'datastack.json')
+    const _ = await this.argsToJsonFile(datastackPath);
```
Yes, good suggestion. I created #19 to cover this. I think there are a couple other cases that should be configureable too.

dcdenu4 commented:
@davemfish That sounds great to me! Thanks for addressing those and I'll let you or James do the honors of merging :)

phargogh commented on src/InvestJob.jsx:
```diff
@@ -38,32 +39,48 @@ if (process.env.GDAL_DATA) {
+const LOGLEVELMAP = {
+  'DEBUG':   '--debug',
+  'INFO':    '-vvv',
+  'WARNING': '-vv',
+  'ERROR':   '-v',
+}
+
 export class InvestJob extends React.Component {
+  /** This component and it's children render all the visible parts of the app.
+  *
+  * This component's state includes all the data needed to represent one invest
+  * job.
+  */
+
   constructor(props) {
     super(props);
 
     this.state = {
-      sessionID: defaultSessionID(''),
+      sessionID: null,                 // modelName + workspace.directory + workspace.suffix
```
Resolved in another conversation on this PR: https://github.com/natcap/invest-gui/pull/16#discussion_r413196264

phargogh commented:
Sounds great @davemfish!
